refs

coot opened the pull request:
* React.DOM.Props.ref - set ref property (string)
* React.readRef - read ref through Foreign

String refs are not yet deprecated, but they might be at some point and React-15.6.1 docs do not event mention them. Nevertheless we should support them.

I'd be keen to change `readRef` type to 
```
readRefs :: String -> Foreign -> Eff (refs :: ReactRef (read :: Read | access) | eff) (Either MultipleErrors Node)
```
i.e. 
 * run `runExcept`
 * use Foreign directly

This way it will be also use able if one sets a ref using a callback. Then on could pass `toForeigh this` and get back the ref.  Setting callback refs is not included in the patch but I'd be happy to find a solution for that - maybe with a more fresh mind in the morning :)

paf31 commented:
I think this means adding a dependency on `foreign` which comes with a lot of dependencies, whereas `react` right now has almost none. I'd prefer to avoid the dependency if possible.

coot commented:
The other option is to use `Refs` instead of `Foriegn` and to provide safe way of writing a `ref` to `Refs` and reading it from there (as a `Maybe`).  Then we could also support callback arguments for the ref property which would write to `Refs`.

coot commented:
* now it only introduces `purescript-dom` dependency, which anyway will be in any project
* I think we could be less conservative and use `HTMLElement` instead of `Node` type.  Storing an attribute node (which is a Node but not Element), is possible but not something usable.
* we could be more conservative on effects that we allow in `refCb`, for example `(refs :: ReactRef (write :: Write | access), state :: ReactState () | eff)`, there is no point on reading the state, and writting it will actually raise an exception

ethul commented:
I think this looks good. I want to take a bit closer look when I am at my
computer. However, one question is regarding readRefImpl. Wouldn't this
return undefined instead of null? There is a purescript-undefinable module.

Thanks for working on this!

On Sat, Jul 8, 2017 at 08:17 Coot <notifications@github.com> wrote:

>
>    - now it only introduces purescript-dom dependency, which anyway will
>    be in any project
>    - I think we could be less conservative and use HTMLElement instead of
>    Node type. Storing an attribute node (which is a Node but not
>    Element), is possible but not something usable.
>    - we could be more conservative on effects that we allow in refCb, for
>    example (refs :: ReactRef (write :: Write | access), state ::
>    ReactState () | eff), there is no point on reading the state, and
>    writting it will actually raise an exception
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/purescript-contrib/purescript-react/pull/100#issuecomment-313852609>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAVYy5w-4Ytto8S6F_NCE57Tl29AHAebks5sL3NcgaJpZM4OHRBs>
> .
>

coot commented:
purescript-nullable is using `== null` to check if something is `null` or not and `undefined == null` returns `true`.  The`refCb` is called with `null`, so it might be set to null. [Undefinable](https://github.com/ethul/purescript-undefinable/blob/master/src/Data/Undefinable.js#L6) is using `=== undefined`  so this will not detect a null.  So `Nullable` looks safer to me.

ethul commented:
Gotcha. Thanks for confirming

On Sat, Jul 8, 2017 at 08:49 Coot <notifications@github.com> wrote:

> purescript-nullable is using == null to check if something is null or not
> and undefined == null returns true. TherefCb is called with null, so it
> might be set to null. Undefinable
> <https://github.com/ethul/purescript-undefinable/blob/master/src/Data/Undefinable.js#L6>
> is using === undefined so this will not detect a null. So Nullable looks
> safer to me.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/purescript-contrib/purescript-react/pull/100#issuecomment-313854062>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAVYy4YUP6Ua2sv42-vGpmf6OLp82uA_ks5sL3rhgaJpZM4OHRBs>
> .
>

paf31 commented:
Unfortunately, `dom` still pulls in quite a few dependencies. I'd really like to avoid adding more dependencies here if we can - I really like that all you need is `prelude` and `eff`.

I'll have a think about some options.

coot commented:
We could make refs polymorphic in the type. That would leave choice weather use `Foreign` or `HTMLElement` or a some more specific type like `HTMLInputElement` (`Foreign` seems the safest and most generic type though.)

Since this library is oriented towards the web, a requirement of `purescript-dom` or even `purescript-foreign` looks really minimal to me; `purescript-react-dom` already requires `purescript-dom` (and thus `purescript-foreign`).

ethul commented:
After thinking about this further. I agree with @paf31. I think if we can avoid depending on `purescript-dom`, it would be ideal. For example, I don't think we'd necessarily need `purescript-dom` in a react native app.

ethul commented on src/React/DOM/Props.purs:
````diff
@@ -297,6 +301,19 @@ radioGroup = unsafeMkProps "radioGroup"
+-- | ```
+refCb
+  :: forall access eff
+   . (Node -> Eff (refs :: ReactRefs (write :: Write | access) | eff) Unit)
````
Just a quick note, would it make sense for this to be `Nullable Node`?
https://facebook.github.io/react/docs/refs-and-the-dom.html#caveats

coot replied:
Yes

ethul commented:
I think it would be very useful to merge in this PR. However, I do think it would be best to avoid the DOM dependency. Would it make sense to replace `Node` with just a `ref`? E.g.,

```purescript
ref :: forall access eff ref. (Nullable ref -> Eff (refs :: ReactRefs (write :: Write | access) | eff) Unit)
```

paf31 commented:
But wouldn't the user be able to use such a function unsafely? 

Do we even use the `ref` type anywhere? What about just creating a new `foreign data` type, and then the user can coerce it if they want to pull in `dom`.

ethul commented:
Indeed it could be used in an unsafe way. I think a new foreign type would
work.

On Fri, Sep 1, 2017 at 20:53 Phil Freeman <notifications@github.com> wrote:

> But wouldn't the user be able to use such a function unsafely?
>
> Do we even use the ref type anywhere? What about just creating a new foreign
> data type, and then the user can coerce it if they want to pull in dom.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/purescript-contrib/purescript-react/pull/100#issuecomment-326711211>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAVYy1AKFKj1l_XrG5B5Mrw-QemS5gkyks5seKb7gaJpZM4OHRBs>
> .
>

paf31 commented on src/React.purs:
```diff
@@ -294,6 +300,26 @@ foreign import getRefs :: forall props state access eff.
   ReactThis props state ->
   Eff (refs :: ReactRefs (read :: Read | access) | eff) Refs
 
+foreign import data Ref :: Type
```
Could you please add a comment here?

paf31 commented on src/React/DOM/Props.purs:
````diff
@@ -297,6 +300,19 @@ radioGroup = unsafeMkProps "radioGroup"
+-- | ``` purescrript
+-- | div [ refCb (writeRef this "inputEl") ] [...]
+-- | ```
+refCb
````
I don't like this name so much, what about something like `withRef`?

coot replied:
I actually I don't like it either, `withRef` sounds much better :)

paf31 commented on src/React.purs:
```diff
@@ -294,6 +300,26 @@ foreign import getRefs :: forall props state access eff.
+  String ->
+  Eff (refs :: ReactRefs (read :: Read | access) | eff) (Nullable Ref)
+
+readRef :: forall props state access eff.
```
Could you please add comments to these as well?

coot replied:
`Ref` type should also be documented, and advising the user to use `unsafeCoerce` make me feel I am doing something wrong, so either we provide a third library with `Node -> Ref` and `Ref -> Node` safe functions (`purescript-react-refs`) or we can include that in `purescript-react-dom`, which already depends on `purescript-dom`. 
 There are pros and cons of both, but either will provide a safe way for the basic operation.

By the way: using zypher & webpack (with uglify plugin) on [purescript-isomorphic-react-example](https://github.com/coot/purescript-isomorphic-react-examples), which is pulling DOM  for very basic stuff (access to location, history, search the mounting point in the DOM), pulls ~2.3KB of data from `purescript-dom`, which is ~0.5% of whole bundle (~400KB, which gzips to 94KB). zypher, to some extend can remove things from foreign modules that's why the overall code from DOM that lands in the bundle is pretty minimal.

ethul replied:
In my opinion, I think `purescript-react-dom` would be a nice spot for functions that convert between `Ref` and `Node`. Similarly, in a `purescript-react-native` library, there could be functions that convert between `Ref` and something native.

ethul commented on src/React/DOM/Props.purs:
````diff
@@ -297,6 +300,19 @@ radioGroup = unsafeMkProps "radioGroup"
+-- | ```
+refCb
+  :: forall access eff
+   . (Ref -> Eff (refs :: ReactRefs (write :: Write | access) | eff) Unit)
````
Just as a note, I think this may need to be

```purescript
(Nullable Ref -> Eff (refs :: ReactRefs (write :: Write | access) | eff) Unit)
```

coot replied:
From react docs:

React will call the ref callback with the DOM element when the component mounts, and call it with null when it unmounts.

And this is done to avoid memory leaks (DOM references used to leak in old IEs).  If we write the ref only if it's not `null`, you'll prevent the original mechanism to prevent memory leaks.

So if we change this type signature (which I do like, since things are explicitly typed then) then we also need to change `writeRef :: forall p s. ReactThis p s -> String -> Nullable Ref -> Eff ...`

ethul replied:
Agreed. I think it makes sense to write a `Nullable Ref`.

coot commented:
I added comments and renamed the function `withRefs`. Another PR will go to purescritp-react-dom` which adds `refToNode` which is mentioned in the comments.

ethul commented on src/React/DOM/Props.purs:
````diff
@@ -297,6 +300,19 @@ radioGroup = unsafeMkProps "radioGroup"
+-- | ```
+withRef
+  :: forall access eff
+   . (Ref -> Eff (refs :: ReactRefs (write :: Write | access) | eff) Unit)
````
Looks great! Quick question did we want this to be `Nullable Ref`? Or are we thinking that the possibility the ref is null would be handled elsewhere.
https://facebook.github.io/react/docs/refs-and-the-dom.html#caveats

coot replied:
Good point.  Having here `Nullable Ref` will be more explicit, so I prefer this. Then in the `React.DOM.refToNode` we will not need to use `unsafeCoerce`

coot replied:
We will only need to add a comment, that if one is attaching the ref to `ReactThis` one should do that using the `Nullable Ref` rather than `Ref` itself.

ethul replied:
Great! Thanks!

ethul commented on src/React/DOM/Props.purs:
````diff
@@ -297,6 +301,19 @@ radioGroup = unsafeMkProps "radioGroup"
+ref = unsafeMkProps "ref"
+
+-- | You can use `writeRef` to store a reference on `Refs`.
+-- | ``` purescrript
````
Quick note: should be `purescript`

ethul commented:
Looking great! @paf31 Do you have any other comments for this one?
